### PR TITLE
Box logic (Initial Proposal)

### LINKED ISSUE #43: Inconsistent behaviour returning from `eval_HP`
In `eval_HP` return status is not monitored yet HP parameter is of intent=OUT.
See https://github.com/ralna/RALFit/blob/2b0dfc8bb9d34e8d1c302842964afa951bcfcacb/libRALFit/src/ral_nlls_internal.f90#L3597

Compiling with `nagfor -nan` will return NANs in HP illustrating the issue and thus breaking.

Issues are:

1. even if `eval_hp_provided=.True.` return `status` should be monitored and allow calls to `dgemv` if appropriate, otherwise it should abort with ERROR status -which at me moment is not possible-.

2. current behaviour is to return HP=0 and this is not correct.

```{Fortran}
subroutine calculate_sHs( n, m, s, params) ! change to
subroutine calculate_sHs( n, m, s, params, status)

       If (params%eval_hp_provided) Then
          Call params%eval_hp(en_data=en_data,objhprd=objhprd,                 &
            objhprdh=objhprdh,iuser=iuser,ruser=ruser,status=status,n=n,       &
            m=params%m,x=params%x,y=s(1:n),hp=params%tenj%hs,                  &
            params=params%parent_params)
          If (status/=0) Then
!           Error eval_hp returned non-zero status, propagate back!
            Go To 100
          End If
        Else
          Do ii = 1, params%m
            params%tenj%h(1:n,1:n) = params%hi(1:n,1:n,ii)
            params%tenj%hs(1:n,ii) = 0.0_wp
            Call dgemv(trans='N',m=n,n=n,alpha=1.0_wp,a=params%tenj%h(1,1),    &
              lda=n,x=s,incx=1,beta=0.0_wp,y=params%tenj%hs(1,ii),incy=1)
          End Do
        End If
        Call dgemv(trans='T',m=n,n=params%m,alpha=1.0_wp,                      &
          a=params%tenj%hs(1,1),lda=n,x=s,incx=1,beta=0.0_wp,                  &
          y=params%tenj%sths(1),incy=1)
100     Continue
```

### LINKED ISSUE #44: `norm_S_d` undefined under some option combinations
This is an analogous issue to the undefined `nord_2_d` issue in `calculate_step` and is **solved in the `box-logic` branch**.

Symptom: `norm_S_d` has intent OUT yet it can be undefined under certain conditions.

https://github.com/ralna/RALFit/blob/2b0dfc8bb9d34e8d1c302842964afa951bcfcacb/libRALFit/src/ral_nlls_internal.f90#L1052
Change to
```{Fortran}
    norm_2_d = 1.0e10_wp ! set norm_2_d so that it can't be undefined later
    norm_S_d = 1.0e10_wp ! set norm_S_d so that it can't be undefined later
```

### LINKED ISSUE #45: Wrong count of Hessian evaluations
Symptom: Routine `get_Hi` does not account for the calls made to `eval_HF`. **This issue is solved in the `box-logic` branch**.

Solution: Update `inform%h_eval`.

```{Fortran}
       if ( present(weights) ) then
          call eval_HF(inform%external_return, n, m, X, weights(1:m)*ei, Hi, params)
       else
          call eval_HF(inform%external_return, n, m, X, ei, Hi, params)
       end if
       inform%h_eval = inform%h_eval + 1
       If (inform%external_return/=0) Then
         inform%status = NLLS_ERROR_FROM_EXTERNAL
         inform%external_name = "eval_HF"
       End iF
     end subroutine get_Hi
```

### LINKED ISSUE #57: Setup workspace routines clear inform derive type
@tyronerees The routines `setup_workspace*` have `inform` with `intent(Out)` resetting to default values each time they are called, this is not the desired behaviour under the new fallback strategy, where some workspaces are not yet available and `setup_workspace*` is called and hence clears `inform`, rationales is to keep `NLLS_SOLVE` with `intent(Out)` for `inform` and all other routines change to `intent(InOut)`.

Affects
```
setup_workspace_aint_tr.f90
setup_workspace_all_eig_symm.f90
setup_workspace_calculate_step.f90
setup_workspace_dogleg.f90
setup_workspace_evaluate_model.f90
setup_workspace_generate_scaling.f90
setup_workspace_max_eig.f90
setup_workspace_min_eig_symm.f90
setup_workspace_more_sorensen.f90
setup_workspace_regularization_solver.f90
setup_workspace_solve_general.f90
setup_workspace_solve_lls.f90
setup_workspace_solve_newton_tensor.f90
setup_workspaces.f90
```

### LINKED ISSUE #65: C <-> Box Logic
- [ ] Set up interface
- [ ] Add simple test
- [ ] Update documentation

### LINKED ISSUE #72: Tensor-Newton subproblem non-zero exit `tensor_inform%status`
If Tensor-Newton sub-problem fails (solved via recursive call to `nlls_iterate` the `tensor_inform%status` is not propagated to `inform%status`, actually, tensor_inform%status is only used to break the iteration loop and is never propagated! see [here](https://github.com/ralna/RALFit/blob/dece9fc8bb5fd8ed3bafeb0238a2171a2af06f0b/libRALFit/src/ral_nlls_internal.f90#L3656)

Proposed fix is to add a check and propagate on non-zero exit status after
calling `nlls_finalize`.

The proposed change aligns with the current behaviour of all other sub-problems in the `allow_fallback` scheme.

**NOTE: This change induces hard exit (no_progress_x error) of (outer) `calculate_step` due to (inner) `calculate_step` non-zero exit status.**

### LINKED ISSUE #76: Removal of Automatic Arrays
# Remove automatic arrays from source

Code has TWO type of triggers for mallocs:

## Type one
Calls of the form
```
Subroutine a(n)
      Integer, Intent(In) :: n
      Real(Kind=wp) :: vector(n)
...
```
Trigger `malloc`s  and should be removed. A fix could be to propagate workspaces...
```
Subroutine a(n, vector)
      Integer, Intent(In) :: n
      Real(Kind=wp), Intent(Inout) :: vector(n)
...
```

Outstanding RALFit `malloc`s are
 * GET_HI automatic array, EI [checkCoding-E-48]
Ringfences GALAHAD code
 * `DRQS_SOLVE` automatic array, `C_SCALE` [checkCoding-E-48]
 * `DRQS_SOLVE` automatic array, `H_SCALE` [checkCoding-E-48]
 * `DTRS_SOLVE` automatic array, `C_SCALE` [checkCoding-E-48]
 * `DTRS_SOLVE` automatic array, `H_SCALE` [checkCoding-E-48]

Other `malloc`s are related to calling routines with expresion of the form `-vector(1:n)`
```{Fortran}
Call routine(n, -1.0_wp * vector(1:n))
```
`malloc`s for the result of `-1.0_wp * vector(1:n)`.

## Type two
Dummy arguments assumed shape/size translation

`j(*)` is addressed as a contiguous block of memory.

`j(:)` is a structure which varies by compiler

The `j(:)` memory metadata has to be traversed and flattened to contiguity before passing to `j(*)`, which is why the temp is created.

Giving `j(:)` the `Contiguous` attribute should help, that would require it to always be passed 'sensible' Allocatable stuff without strides, and not Pointers.

# Proposal for `get_Hi`
Issue partially fixed by #78, still need to address automatic array `ei(m)`
```
     Recursive subroutine get_Hi(n, m, X, params, i, Hi, eval_HF, inform, ei, weights)
       Implicit None
       integer, intent(in) :: n, m
       real(wp), intent(in) :: X(:)
       class( params_base_type ), intent(inout) :: params
       integer, intent(in) :: i
       real(wp), intent(out) :: Hi(:,:)
       procedure( eval_hf_type ) :: eval_HF
       type( nlls_inform ), intent( inout ) :: inform
       real( wp ), dimension( m ), intent( in ), optional :: weights
       real( wp ), Intent(Inout) :: ei( m ) ! <--- Avoid automatic arrays, trigger malloc on demand

       ei(:) = 0.0_wp
       If ( present(weights) ) Then
        ei(i) = weights(i)
       Else
        ei(i) = 1.0_wp
       End If

!      call eval_HF(inform%external_return, n, m, X, weights(1:m)*ei, Hi, params) !<---- avoid trigger to malloc 
!      for the storage of weights(:)*ei wich is double waste of space and work.
!      We should inform the user NOT do to the HP product blindly, first check non-zero entries of ei, which 
!      by construction here has only ONE non-zero.

       call eval_HF(inform%external_return, n, m, X, ei, Hi, params)

       inform%h_eval = inform%h_eval + 1
       If (inform%external_return/=0) Then
         inform%status = NLLS_ERROR_EVALUATION
         inform%external_name = "eval_HF"
         inform%external_return = 2513
       End iF
     end subroutine get_Hi
```

### LINKED ISSUE #79: Solution for underdetermined systems (n>m)
Return a solution in the case of **underdetermined** systems.

Currently it will return an error, it would me more useful to estimate a solution.
(Some CUTEst problems are underdetermined)

### LINKED ISSUE #90: Use blas where appropriate
Using BLAS would be more efficient than looping through entries, especially for large arrays.  

Lines that might benefit include:

https://github.com/ralna/RALFit/blob/cea1ab2cb7108b2ad75e41c70e6b27c5c63924ee/libRALFit/src/ral_nlls_internal.f90#L1812-L1816

https://github.com/ralna/RALFit/blob/cea1ab2cb7108b2ad75e41c70e6b27c5c63924ee/libRALFit/src/ral_nlls_internal.f90#L1967-L1973

https://github.com/ralna/RALFit/blob/cea1ab2cb7108b2ad75e41c70e6b27c5c63924ee/libRALFit/src/ral_nlls_internal.f90#L3036-L3040

https://github.com/ralna/RALFit/blob/cea1ab2cb7108b2ad75e41c70e6b27c5c63924ee/libRALFit/src/ral_nlls_internal.f90#L3648-L3650

https://github.com/ralna/RALFit/blob/cea1ab2cb7108b2ad75e41c70e6b27c5c63924ee/libRALFit/src/ral_nlls_internal.f90#L4104-L4107

https://github.com/ralna/RALFit/blob/cea1ab2cb7108b2ad75e41c70e6b27c5c63924ee/libRALFit/src/ral_nlls_internal.f90#L4113-L4115

https://github.com/ralna/RALFit/blob/cea1ab2cb7108b2ad75e41c70e6b27c5c63924ee/libRALFit/src/ral_nlls_internal.f90#L4183-L4185

### PR DESCRIPTION
**Pull request by talassio**
- [x] STP (test deck+coverage)

This PR:
fixes #43 
fixes #44 
fixes #45 
fixes #57 
fixes #65 
fixes #72 
fixes #76
fixes #79 
fixes #90

### CODE REVIEW

**Comment by tyronerees** [comment]
I don't know what the issue would be, but just to note that the python test fails on my system with these changes:
```
 ===========================
 =--Testing the main file--=
 ===========================
 *** All (main) tests passed successfully! ***
 =============================
 =--Testing the subroutines--=
 =============================
 *** All (helper) tests passed successfully! ***
  
 **************************************
 *** All tests passed successfully! ***
 **************************************
Note: The following floating-point exceptions are signalling: IEEE_UNDERFLOW_FLAG IEEE_DENORMAL
STOP 0
** C test passed successfully **
3c3
< The cost function is 0.000000 at the minimum, and ||J^Tr||=2.247131e+00
---
> The cost function is 2.247131 at the minimum, and ||J^Tr||=4.097987e-06
[Python test]: Something is wrong with the diff
```
Interesting that the C and Fortran tests passed...hopefully this will change naturally when python is properly hooked in

File: libRALFit/example/Fortran/CMakeLists.txt [hunk 1]
```diff
@@ -9,5 +9,11 @@ target_link_libraries(nlls_example2 ral_nlls)
 add_executable (lanczos Lanczos.f90)
 target_link_libraries(lanczos ral_nlls)
 
+add_executable (LanczosBox LanczosBox.f90)
+target_link_libraries(LanczosBox ral_nlls)
+
+add_executable (stp_x0sol stp_x0sol.f90)
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/example/Fortran/CMakeLists.txt]: Yes, it only need to be a test.
Check for 
 * `inform%iter==0`
 * `inform%nf == 1`
 * `inform%ng == 1`

File: libRALFit/src/ral_nlls_workspaces.f90 [hunk 2]
```diff
@@ -307,6 +317,56 @@ module ral_nlls_workspaces
...
 
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!!! B O X   B O U N D   O P T I O N S !!!
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+!    Memory size for the non-monotone linesearch
+     Integer :: box_nFref_max = 4
+!    Kanzow sufficient decrease ratio (eq 25) Kanzow 2004
+     Real(Kind=wp) :: box_gamma = 0.99995_wp
+     Real(Kind=wp) :: box_decmin = 2.0_wp * 1.0e-16_wp ! macheps
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: Yes, in will be change to `eps`.

File: libRALFit/src/ral_nlls_workspaces.f90 [hunk 3]
```diff
@@ -307,6 +317,56 @@ module ral_nlls_workspaces
...
+!    Magic number to consider box bound (+/-) infinity
+     Real(Kind=wp) :: box_bigbnd = 1.0e20_wp
+!    Wolfe descent condition (0<\sigma1<1/2), curvature condition (0<\sigma2)
+     Real(Kind=wp) :: box_wolfe_descent = 1.0e-4_wp
+     Real(Kind=wp) :: box_wolfe_curvature = 0.9_wp
+!    Tolerance to consider projected dir a descent direction
+!    See LS STEP Section 4 p392 Kanzow 2014
+     Real(Kind=wp) :: box_kanzow_power = 2.1_wp
+!    sqrt(mcheps)
+     Real(Kind=wp) :: box_kanzow_descent = 1.0e-8_wp
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: yes, it iwill.

File: libRALFit/src/ral_nlls_workspaces.f90 [hunk 4]
```diff
@@ -307,6 +317,56 @@ module ral_nlls_workspaces
...
+!    sqrt(mcheps)
+     Real(Kind=wp) :: box_quad_model_descent = 1.0e-8_wp
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: Yes, this also will be changed

File: libRALFit/src/ral_nlls_workspaces.f90 [hunk 5]
```diff
@@ -382,23 +448,24 @@ module ral_nlls_workspaces
 
 !  vector of smallest singular values
 
-     real(wp), allocatable :: smallest_sv(:)
+!    real(wp), allocatable :: smallest_sv(:)
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: ;-)

File: libRALFit/src/ral_nlls_workspaces.f90 [hunk 6]
```diff
@@ -382,23 +448,24 @@ module ral_nlls_workspaces
...
 
 !  vector of largest singular values
 
-     real(wp), allocatable :: largest_sv(:)
+!    real(wp), allocatable :: largest_sv(:)
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: ;-)

File: libRALFit/src/ral_nlls_workspaces.f90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: changed

File: libRALFit/src/ral_nlls_workspaces.f90 (hunk 3 above)
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: changed

File: libRALFit/src/ral_nlls_workspaces.f90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: changed

File: libRALFit/src/ral_nlls_workspaces.f90 [hunk 7]
```diff
@@ -307,6 +317,56 @@ module ral_nlls_workspaces
...
+!    to scale_alpha*alpha_k-1
+     Integer       :: box_quad_match = 1
+!    Initial step scale (if quad_i >= box_quad_i)
+     Real(Kind=wp) :: box_alpha_scale = 1.0_wp
+!    Scaling factor to use when updating Delta from LS/PG step
+     Real(Kind=wp) :: box_Delta_scale = 2.0_wp
+     Real(Kind=wp) :: box_tau_wolfe = 0.3_wp
+     Real(Kind=wp) :: box_tau_tr_step = 0.3_wp
+     Integer       :: box_ls_step_maxit = 20
+!    LS type: 1 => Dennis-Schnable; 2 => Hager-Zhang
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: changed

**Comment by tyronerees** [comment]
This merge adds the memory allocation corrections that are in master.  There were merge conflicts, but (hopefully!) they were resolved straightforwardly

File: libRALFit/src/ral_nlls_internal.f90 [hunk 8]
```diff
@@ -2148,7 +2435,10 @@ subroutine more_sorensen_noew(A,v,n,m,Delta,d,nd,options,inform,w)
                 Write(rec(1), Fmt=6020) sigma_l
                 Call printmsg(5,.False.,options,1,rec)
               End If
-              dHd = dot_product(d, matmul(w%AplusSigma,d))
+             !dHd = dot_product(d, matmul(w%AplusSigma,d))
+             !reuse w%q to store intermediary MV product
+              w%q(:) = matmul(w%AplusSigma,d)
```
Comment (talassio) [inline libRALFit/src/ral_nlls_internal.f90]: re-use array w%q and estimate the <d,d>_H in two steps

File: libRALFit/src/ral_nlls_internal.f90 [hunk 9]
```diff
@@ -3245,7 +3552,8 @@ subroutine solve_spd(A,b,LtL,x,n,inform)
        inform%status = 0
        ! wrapper for the lapack subroutine dposv
        LtL(1:n,1:n) = A(1:n,1:n)
-       x(1:n) = b(1:n)
+       ! NOTE: b is inverted!
```
Comment (talassio) [inline libRALFit/src/ral_nlls_internal.f90]: b is actually -b so it solves Ax=-b, this also applies to solve_general and solve_spd_nocopy (STP also extended), this avoid auto allocs in all the places that calls solve_xxx(..., A, -v, ...)

File: libRALFit/src/ral_nlls_internal.f90 [hunk 10]
```diff
@@ -4118,7 +4118,7 @@ Recursive subroutine evaltensor_J(status, n, m, s, J, params)
        real(wp), dimension(*), intent(in)    :: s
        real(wp), dimension(*), intent(out)   :: J
        class( params_base_type ), intent(inout) :: params
-       integer :: ii, jj
```
Comment (talassio) [inline libRALFit/src/ral_nlls_internal.f90]: @tyronerees this is the proposal for `evaltensor_J`to address slices inside assumed-shape arrays, can you review it? just to make sure everything is right.

File: libRALFit/src/ral_nlls_internal.f90 [hunk 11]
```diff
@@ -4064,12 +4066,20 @@ Recursive subroutine evaltensor_f(status, n, m, s, f, params)
           call calculate_sHs(n,m, s, params)
 
           ! put them all together for the first 1:m terms of f
-          f(1:params%m) = params%f(1:params%m) + params%tenJ%Js(1:params%m) +  &
```
Comment (talassio) [inline libRALFit/src/ral_nlls_internal.f90]: @tyronerees this is the proposal for `evaltensor_f`to address slices inside assumed-shape arrays, can you review it? just to make sure everything is right.

File: libRALFit/src/ral_nlls_internal.f90 (hunk 9 above)
Comment (talassio) [reply libRALFit/src/ral_nlls_internal.f90]: @tyronerees suggested a name change. e.g. in `solve_spd` to account for the inverting of `b` in `Ax=-b`.

File: libRALFit/src/ral_nlls_dtrs_double.f90 [hunk 12]
```diff
@@ -1560,11 +1574,9 @@ SUBROUTINE DTRS_solve_main( n, radius, f, C, H, X, control, inform )
 !  record all of the estimates of the optimal lambda
 
         IF ( printd ) THEN
+          MLOC(:) = MAXLOC( lambda_new( : n_lambda ) )
           WRITE( out, "( A, ' lambda_t (', I1, ')', 3ES20.13 )" )              &
-            prefix, MAXLOC( lambda_new( : n_lambda ) ),                        &
-            lambda_new( : MIN( 3, n_lambda ) )
-          IF ( n_lambda > 3 ) WRITE( out, "( A, 13X, 3ES20.13 )" )             &
-            prefix, lambda_new( 4 : MIN( 6, n_lambda ) )
```
Comment (talassio) [inline libRALFit/src/ral_nlls_dtrs_double.f90]: Here `lambda_new` is of length 1:3.

File: libRALFit/src/ral_nlls_dtrs_double.f90 [hunk 13]
```diff
@@ -2932,11 +2957,13 @@ SUBROUTINE DRQS_solve_main( n, p, sigma, f, C, H, X, control, inform )
...
         IF ( printd ) THEN
+          MLOC(:) = MAXLOC( lambda_new( : n_lambda ) )
           WRITE( out, "( A, ' lambda_t (', I1, ')', 3ES20.13 )" )              &
-            prefix, MAXLOC( lambda_new( : n_lambda ) ),                        &
-            lambda_new( : MIN( 3, n_lambda ) )
+            prefix, MLOC(1), lambda_new( : MIN( 3, n_lambda ) )
           IF ( n_lambda > 3 ) WRITE( out, "( A, 13X, 3ES20.13 )" )             &
             prefix, lambda_new( 4 : MIN( 6, n_lambda ) )
+          IF ( n_lambda > 6 ) WRITE( out, "( A, 13X, 3ES20.13 )" )             &
+            prefix, lambda_new( 7 : MIN( 9, n_lambda ) )
```
Comment (talassio) [inline libRALFit/src/ral_nlls_dtrs_double.f90]: Here `lambda_new` is of length 1:6.

File: libRALFit/src/ral_nlls_internal.f90 [hunk 14]
```diff
@@ -3928,27 +3949,30 @@ Recursive subroutine solve_newton_tensor(J, f, eval_HF, X, n, m, Delta, &
...
 
+       ! save to params
+       w%tparams%f(1:m) = f(1:m)
+       w%tparams%Delta = Delta
+       w%tparams%J(1:n*m) = J(1:n*m)
+       w%tparams%X(1:n) = X(1:n)
+       w%tparams%eval_HF => eval_HF
+       w%tparams%parent_params => params
+       w%tparams%tenJ => tenJ
+
```
Comment (talassio) [inline libRALFit/src/ral_nlls_internal.f90]: `get_Hi` now has an extra array `ei(m)`, a work array, so it does not require the automatic array equivalent. The idea is to use `w%tparams%f(1:m)` just before it is initialized with `f(1:m)`. In order to exploit this, the initialization must be postponed after the call to `get_Hi` (loop).

File: libRALFit/src/ral_nlls_internal.f90 (hunk 10 above)
Comment (tyronerees) [reply libRALFit/src/ral_nlls_internal.f90:3768]: This looks good to me

**Pull request merged**